File: CollabFilter.scala
import scala.collection.mutable.HashMap;

class CollabFilter() {
    val map = HashMap[String, HashMap[String, Double]]();

    def setRating(user: String, film: String, rating: Double) = map.apply(user).put(film, rating);

    def setUser(user: String) = map.put(user, new HashMap[String, Double]);

    def getRating(user: String, film: String) = map.apply(user).apply(film);

    def rate(user: String, film: String): Double = {
        val users = (for (i <- map.iterator if i._1 != user) yield i._1).toArray; //the set of users
        val filmsRated = map.apply(user).keySet; //the set of films rated by user U
        val films = users.iterator.zip(for (o <- users.iterator) yield (map.apply(o).keySet.intersect(filmsRated))).toArray;
        //films is an Iterator[(String, Set[String])] essentially, it iterates over the pairs (user O, set of shared films)
        //shared with user U by user O
        val top = for (o <- films.iterator) yield o._2.iterator.map((s: String) => map.apply(user).apply(s) * map.apply(o._1).apply(s)).sum;
        val left = for (o <- films.iterator) yield o._2.iterator.map((s: String) => map.apply(user).apply(s) * map.apply(user).apply(s).asInstanceOf[Double]).sum;
        val right = for (o <- films.iterator) yield o._2.iterator.map((s: String) => map.apply(o._1).apply(s) * map.apply(o._1).apply(s).asInstanceOf[Double]).sum;
        val bottom = left.zip(right);
        val sim = top.zip(bottom);
        val simils = users.iterator.zip(sim.map((s) => s._1 / (math.sqrt(s._2._1) * math.sqrt(s._2._2)))).toArray;
        val k = 1 / (for (s <- simils.iterator) yield s._2).sum;
        val weighted = simils.iterator.map((p) => p._2 * map.apply(p._1).apply(film)).sum;
        return weighted * k;
    }
}

File: collabfilter.py
from collections import defaultdict
from math import sqrt

class CollabFilter():
    def __init__(self):
        self.cf = defaultdict(dict)

    def setRating(self, user, film, rating):
        self.cf[user][film] = float(rating)

    def setUser(self, user):
        self.cf[user] = {}

    def getRating(self, user, film):
        return self.cf[user][film]

    def rate(self, user, film):
        other_users = set(self.cf.keys()) - {user}
        filmsRated = set(self.cf[user].keys())
        films = {other: set(self.cf[other].keys()) & filmsRated for other in other_users}
        top = {other: sum(self.cf[user][film] * self.cf[other][film] for film in films[other]) for other in other_users}
        left = {other: sum(self.cf[other][film] ** 2 for film in films[other]) for other in other_users}
        right = {other: sum(self.cf[user][film] ** 2 for film in films[other]) for other in other_users}
        similarities = {other: top[other] / (sqrt(left[other]) * sqrt(right[other])) for other in other_users}
        k = 1/ sum(similarities.values())
        return k * sum(similarities[other] * self.cf[other][film] for other in other_users)
